### ISSUE TITLE
Loss is NaN or inf when finetuning Quartznet model

### ISSUE THREAD

**Issue by jonaskratochvil** [NVIDIA/NeMo#613]
Hello, 

I am finetuning the Quartznet 15x5 model on my own dataset. Training seems to be progressing fine for about 2000 steps but then I start getting the **Loss is NaN or inf** warning and I don't receive any predictions from the model from that point onward and WER is 100% for all batches. 
I have checked my audio data and made sure that there are no samples without text output and filtered out audios with duration less than 0.5 seconds but I still observe the same behavior. Could you please troubleshoot where could the problem be?

Thank you,

Jonas

**Comment by okuchaiev** [comment]
Can you try increasing warmup_steps? Something like 16,000 or 8,000.
Also, when you train with amp_mode set to O1 you should be able to fit about 2x larger batch size compared to O0. So use the maximum batch size you can fit per GPU when using O1.

**Comment by ArtemisZGL** [comment]
I met the same problem, and I tried set the warmp step and lr to different value, but did't work. Also, I observed that the problem always happended in the same step(with batch size 1 and 16), pointing to the same sample, but I check that sample, both the audio and text is correct. And I tried to put the sample to the first place in json, the NaN and inf warning did't show in the first step.  
  
I also used the O1 to training, it can train normally to skip the step gradient overflow, but the warning show so frequently.   
  
I guess it's caused by the data, still debugging...

**Comment by cantwbr** [comment]
Hi, I encountered the same problem after upgrading to 0.10.1. I found that the changes commited in [715401b](https://github.com/NVIDIA/NeMo/commit/715401b26dee5cf4a073717670f18df67dd62a4c) seem to undo the changes of pull #262 and its underlying commit [7311077](https://github.com/NVIDIA/NeMo/pull/262/files/73110772ebb36566ee32851f6662523e5bfbaa06). The related discussion can be found in  (the closed) issue [260](https://github.com/NVIDIA/NeMo/issues/260). 

The CTC loss can produce infinities when the inputs can not be aligned to the targets, which might also be caused by the spectrogram augmentation besides poor audio/text data. Instead of ignoring the infinite loss (which should be the desired behavior and which was introduced with pull #262), the execution is now being terminated, if optimization level O0 is used. 

As a quick fix, I changed Line 48 in nemo/collections/asr/losses.py from
`self._criterion = nn.CTCLoss(blank=self._blank, reduction='none')`
to 
`self._criterion = nn.CTCLoss(blank=self._blank, reduction='none', zero_infinity=True)`
which zeros infinite losses.

**Comment by redoctopus** [comment]
Hi, thanks for letting us know! The change was made in response to #532, where we found that `continue`ing on NaN/inf loss would often desync workers and cause a deadlock later in training.

Originally our plan was to get all workers to skip the step, but this proved to be a bit complicated so it was put on the back burner for when we had more bandwidth, hence the change with a termination on NaN/inf when using O0. We're discussing and will be revisiting the NaN/inf loss behavior.

**Comment by redoctopus** [comment]
#637 is merged and should be closer to the desired behavior--the step is skipped if NaN/inf loss is encountered using optimization level O0.
